### ISSUE TITLE
"Fool-proof" API with regards to CoInitializeEx

### ISSUE THREAD

**Issue by roblabla** [microsoft/windows-rs#1169]
I currently have a rather interesting segfault in pure safe code that I believe is due to windows-rs not handling `CoInitialize`/`CoUninitialize` at all. You can find the reproducer [here](https://github.com/roblabla/reproducer-wmi-windows-rs). Note that you may need to run the reproducer multiple time to reproduce the segfaults (I usually get the segfault every 4-5 runs). 

Here's what happens:

- In one thread, I use windows-rs to do some calls to WinRT. This thread has previously been initialized with CoInitialize, but has since been uninitialized, and thus, the "automatic" initialization of the COM subsystem does *not* occur.

- In another thread, I use the `WMI` crate to interact with the WMI. This enforces a call to `CoInitialize` (through the `COMLibrary::new` constructor), and will call `CoUninitialize` when the `WMIConnection` is dropped.

Here's the problem: When the `WMIConnection` gets dropped, it will also drop the `COMLibrary`, which will call `CoUninitialize`. Because no `CoInitialize` is currently active in the WinRT thread, all currently loaded COM DLLs will be unloaded, *including* whatever DLL was loaded by WinRT. So the thread running WinRT code will segfault because the code it's running gets unloaded.

Note that the above reproducer is written in entirely safe code, so there's a safe method that's lying about their safety invariants here, e.g. a method that breaks invariants that some unsafe code somewhere else relies on. Here are the invariants:

- WMI's `ComLibrary` assumes that calling `CoInitialize`/`CoUninitialize` is always safe, and that any methods that relies on having an initialized COM environment will have themselves called `CoInitialize`/`CoUninitialize`

- Meanwhile, I'm not sure what the invariants are expected on the `windows-rs` side. The library doesn't appear to call `CoInitialize` (or `RoInitialize`) anywhere, so it appears to rely on the automatic initialization, but this may be broken by going through a `CoInitialize`/`CoUninitialize` roundtrip on this thread before executing WinRT code.

IMO, windows-rs should have some mechanism to ensure that the `CoInitialize`/`CoUninitialize` (or RoInitialize) is called before doing any winrt calls.

**Comment by kennykerr** [comment]
COM and WinRT APIs (and Win32 for that matter) have differing requirements on COM apartments that make doing this automatically impractical. Some APIs (even COM APIs) don't require the COM runtime to be initialized at all while others require a specific COM apartment model. There isn't really a reliable way to do this automatically - you really need to know the specific APIs you're using and set up the hosting environment accordingly. In most cases, the best thing you can do is ensure that there is an MTA that is running for the duration of the process. The simplest way to ensure that is by calling `CoIncrementMTAUsage` when the process first starts and then never uninitializing COM on any thread. Additional calls to `CoInitialize` paired with `CoUninitialize` are fine (and sometimes necessary when an STA is required) because they will simply act as additional references to the COM runtime and not cause the reference count to hit zero and begin unloading DLLs. 

For WinRT APIs at least, this is what the Windows crate attempts to do on demand but it really is better just to ensure that COM is loaded from the start and is never unloaded because obviously this concession only applies when WinRT types are activated. 

https://github.com/microsoft/windows-rs/blob/7f24270053c0089071a6b421c8eacffea6e915de/src/runtime/factory_cache.rs#L87

**Comment by roblabla** [comment]
Right, but there is still something wrong here, as safe code can cause segfaults - which rust is supposed to avoid, that's the whole premise of safe vs unsafe. I don't think it's sound for windows-rs' WinRT codepaths to assume that COM is already initialized the way it currently does.

Could the `factory` function ensure `CoIncrementMTAUsage` is always called at least once unconditionally? Right now, it's only executed when there are no MTA active (e.g. when `RoGetActivationFactory` returns `CO_E_NOTINITIALIZED`), but this is inherently racy, which I believe is causing the segfault I observed. E.G. `RoGetActivationFactory()` ends successfully, then another thread calls `CoUninitialized`, now the pointer returned by `RoGetActivationFactory` is invalid and any call made through the vtable will actually segfault.

Having something like

```rust
static VAL: once_cell::sync::Lazy<CO_MTA_USAGE_COOKIE> =  Lazy::new(|| {
    let mut cookie = Default::default();
    let _ = CoIncrementMTAUsage(&mut cookie);
    cookie
});

pub fn factory<C: RuntimeName, I: Interface>() -> Result<I> {
    // Make sure the MTA is initialized.
    let _cookie = Lazy::force(&VAL);
```

Would ensure that the MTA is always initialized - and importantly *stays* initialized until the program ends. It would prevent deinitialization, but at least WinRT calls would be safe.

**Comment by kennykerr** [comment]
I think there are a few different issues here. 

One is whether any calls to foreign (non-Rust) code can be considered safe. @rylev and I have discussed this before and he may have some more thoughts on the subject. What constitutes safe code as far as Rust is concerned? Is a crash unsafe? I assume that's what you mean by "segfault". Typically a Windows API may crash the process by inducing an access violation (e.g. dereferencing a null pointer), to use Windows terminology. This is not considered unsafe from the perspective of the Microsoft security team because an access violation immediately terminates the process and does not allow any additional code to execute.  No unwinding occurs. 

Then there is the matter of whether you can truly ensure that COM-related code will always executed in the proper context. That is almost impossible to guarantee at compile time without careful review and testing. You would literally have to wrap every since call to any indirect function (virtual or import) with a run time check for the apartment state (which would be hugely expensive) and even then you could not really guarantee that some code outside your control would not thwart that check. 

Regarding ensuring that MTA is always initialized for WinRT APIs and remains so until the program ends - that's what the Windows crate already does but only for APIs that fail to load otherwise because some APIs require an explicit STA and an MTA would not suffice. Again, being able to predict the hosting environment for each and every API is not practical.

**Comment by roblabla** [comment]
> What constitutes safe code as far as Rust is concerned?

Rust safe/unsafe is only tangantially related to security, but it's not what defines the boundary between the two. The idea is that Safe Code in Rust is that it should not trigger Undefined Behavior. See the [rustonomicon](https://doc.rust-lang.org/nomicon/safe-unsafe-meaning.html):

> The need for all of this separation boils down a single fundamental property of Safe Rust, the soundness property:
> 
> **No matter what, Safe Rust can't cause Undefined Behavior.**

When Safe Rust can trigger Undefined Behavior, it is because some code is unsound: It does not maintain the invariants necessary for Safe Rust to stay UB-free. (This distinction between unsafe and unsound is kinda important: All safe wrappers around unsafe code need to be sound, e.g. maintain their invariants to ensure safe code cannot trigger UB.)

> Is a crash unsafe?

I am not 100% sure here. I believe the only "crash" (in the segfault/access violation sense) that're "allowed" are stack overflow caused by overly large stack usage, because those are fully reproducible and "defined" (someone will correct me if I'm wrong). However, in the case of the crash we have here, I believe it's actually UB: It's caused by a race condition between two threads, which causes one thread to start executing non-executable/unmapped code regions non-deterministically.

> that's what the Windows crate already does but only for APIs that fail to load otherwise because some APIs require an explicit STA and an MTA would not suffice.

Correct me if I'm wrong (I'm fairly new to COM and WinRT), but couldn't the APIs that require a specific apartment type be marked as such in the WinMD? And do you have any specific example of APIs that have those requirements in WinRT? If we knew those APIs beforehand, we could expose constructors that takes a ZST token to ensure the correct `CoInitialize` function was called.

I know that the rest of COM is a mess, but I believe WinRT is supposed to be the modern, "clean", properly specified APIs... And since WinRT is currently exposed as entirely safe code, it's on the library author to ensure it is at least UB-free. If there truly is no way of ensuring this lack of UB, then I believe all the constructors to WinRT types should be marked unsafe, but I really wish we don't get to this point because it'd be a bit of a stepback in usability.

**Comment by kennykerr** [comment]
UB is definitely bad but dereferencing a null pointer is not considered UB on Windows even if Rust seems to think so. Anyway, this is why all Win32 APIs are assumed to be unsafe while WinRT APIs are assumed to be safe: we're just not sure that we can safely assert that Win32 APIs don't allow UB when used incorrectly. We would like to think that WinRT APIs are "safer" due to a stronger type system in play and if UB occurs it is a bug that should ideally be addressed by the API owners. 

Although the metadata could in theory include more information, it does not and is unlikely to ever include such information. In general, WinRT APIs are expected to include checks to ensure they are used correctly. Those checks just often end up with process termination. Anyway, this thread conflates a few different concerns (e.g. WMI and WinRT are very different APIs). Feel free to open new issues to discuss any specific concerns you may still have.

**Comment by bs2kbs2k** [comment]
> dereferencing a null pointer is not considered UB on Windows even if Rust seems to think so.

No matter what Microsoft or anyone for that regard says, Rust's safety is always based on Rust's definition of UB.

**Comment by roblabla** [comment]
> UB is definitely bad but dereferencing a null pointer is not considered UB on Windows even if Rust seems to think so

That doesn't really matter, the Rust Compiler is allowed to assume no null pointer deref happens, and can (and in fact *does*) optimize aggressively with this assumption. This means that what looks like a simple null pointer deref can turn into much nastier issues. That Windows defines what a null pointer deref does doesn't matter. As far as rust is concerned, the only way to rely on this behavior is through a raw assembly call (or FFI). Even raw pointers cannot be soundly dereferenced, as the Rustonomicon says:

> Unlike C, Undefined Behavior is pretty limited in scope in Rust. All the core language cares about is preventing the following things:
> - Dereferencing (using the * operator on) dangling or unaligned pointers (see below)
> [...]
> A reference/pointer is "dangling" if it is null or not all of the bytes it points to are part of the same allocation (so in particular they all have to be part of some allocation).

- https://doc.rust-lang.org/nomicon/what-unsafe-does.html

----------

> Anyway, this thread conflates a few different concerns (e.g. WMI and WinRT are very different APIs). Feel free to open new issues to discuss any specific concerns you may still have.

This thread conflates them because today, right now, I can mix two crates that both have a safe API and cause UB. That's just not supposed to happen. It does not matter that the APIs are different, really. Basically, one crate has broken the others' assumptions, and we should work to resolve this so the assumptions on both side can be made coherent.

As I said earlier, the assumptions made on both sides are this:

- WMI's `ComLibrary` assumes that calling `CoInitialize`/`CoUninitialize` is always safe, and that any other libraries that relies on having an initialized COM environment will have themselves called `CoInitialize`/`CoUninitialize`

- `windows-rs` basically assumes it is either the first thing to do COM (allowing itself to call `CoIncrementMTAUsage`) or that other things doing COM will never call `CoUninitialize` while it is running.

My understanding from the conversation is that you think that `WMI` is "wrong" in exposing `CoInitializeEx`/`CoUninitialize` to safe code. I opened an issue to them: https://github.com/ohadravid/wmi-rs/issues/39

**Comment by kennykerr** [comment]
Just to be clear, I'm not trying to say that Rust's notion of safety with respect to pointer dereferencing is somehow wrong or inapplicable on Windows. I'm just pointing out what happens inside Windows, not inside Rust code. I am committed to honoring Rust's definition of safety to the best of my ability in the implementation of the Windows crate. The Windows operating system just has a very different understanding of some of these concepts but they need not be in conflict - we just need to figure out how best to make them work together and for that I am grateful for all your input and feedback. 

So then, if you call some safe Rust code that ends up calling some Windows API that ends up calling [TerminateProcess](https://docs.microsoft.com/en-us/windows/win32/api/processthreadsapi/nf-processthreadsapi-terminateprocess) or [abort](https://docs.microsoft.com/en-us/cpp/cpp/program-termination?view=msvc-160) in C++ resulting in a [fast fail](https://docs.microsoft.com/en-us/cpp/intrinsics/fastfail?view=msvc-160) would you consider that a violation of Rust's definition of safety? If that is the case, then there is very little Rust code that could be considered safe including much of the standard library. So hopefully that's not the case. 😉 

Regarding [CoUninitialize](https://docs.microsoft.com/en-us/windows/win32/api/combaseapi/nf-combaseapi-couninitialize), from a Windows perspective that is a very unsafe function call to make. There are legitimate reasons for using it but they certainly fall into the expert user category.

**Comment by kennykerr** [comment]
This is perhaps a little off-topic but I'll just drop this here if it helps: Regarding [CoIncrementMTAUsage](https://docs.microsoft.com/en-us/windows/win32/api/combaseapi/nf-combaseapi-coincrementmtausage), in an ideal world the [CRT](https://docs.microsoft.com/en-us/cpp/c-runtime-library/crt-library-features?view=msvc-160) calls this function early in the startup process (before `WinMain` or its language equivalent) and ensures that the COM runtime is active for the duration of the process thus eliminating any race with different APIs and libraries. This is in fact what the Visual C++ CRT does in some flavors or configurations but is not at present the universal behavior. If there were a way to do this in Rust without relying on the CRT to do so that would be great.

**Comment by roblabla** [comment]
> So then, if you call some safe Rust code that ends up calling some Windows API that ends up calling TerminateProcess or abort in C++ resulting in a fast fail would you consider that a violation of Rust's definition of safety? 

That's totally fine. `abort` is well-defined to simply stop the current execution, and Rust in fact has a safe function that does this in the standard library as well ([`std::process::abort`](https://doc.rust-lang.org/stable/std/process/fn.abort.html) and [`std::process::exit`](https://doc.rust-lang.org/stable/std/process/fn.exit.html)). Reimplementing them by calling `TerminateProcess` is also fine. There's no UB here, it's just calling a C function that has the well-defined behavior of stopping the process.

As far as our specific crash is concerned, I guess the argument could be made that unmapping a DLL whose code is being executed by another thread is not UB... All I can say is that it feels UB to me.

FWIW, I'm not an expert on safe/unsafe - I'm basing my answers on what I remember from the rustonomicon and the [UCG WG discussions](https://github.com/rust-lang/unsafe-code-guidelines). If someone is able to answer this question more firmly, it's the people from the unsafe-code-guidelines WG.

> Regarding CoUninitialize, from a Windows perspective that is a very unsafe function call to make. There are legitimate reasons for using it but they certainly fall into the expert user category.

Alright, then I guess I'll try to work with the wmi-rs crate authors to see what we can do to avoid using that function in safe code.

**Comment by kennykerr** [comment]
> That's totally fine. abort is well-defined 

Great, then I'd just confirm that the crash you're seeing is in fact that same behavior. You should be able to tell by hooking up a system debugger like WinDbg. In all likelihood it is. Windows goes to great lengths to fast fail when things go wrong precisely because this is well-defined.

**Comment by riverar** [comment]
The provided example is panicking on the None unwrap and not a direct result of anything in the crate or Windows. Am I missing something in the repro case?

```
thread '<unnamed>' panicked at 'called `Option::unwrap()` on a `None` value', src\main.rs:21:39
```

```diff
fn inner_wrt() {
    let before = std::time::Instant::now();
+    let pm = Windows::Management::Deployment::PackageManager::new().ok().unwrap();
    let iter = pm.FindPackages().ok().unwrap();
// ...
```

**Comment by roblabla** [comment]
Hmm. The .ok() is unnecessary here. I pushed an extra commit that removes it, that should show the underlying error.

FWIW, the reproducer is not 100% reproducible. I have to run it multiple times before it actually triggers an ACCESS_VIOLATION. It will sometimes succeed, and sometimes fail with an error.

**Comment by kennykerr** [comment]
For what its worth, turns out the CRT option that I was alluding to has already shipped:

exe_initialize_mta.lib

https://docs.microsoft.com/en-us/cpp/c-runtime-library/link-options?view=msvc-160

That would make these sorts of problems go away.

**Comment by roblabla** [comment]
> I'd just confirm that the crash you're seeing is in fact that same behavior.

Also, I disagree with this. We have a race condition triggering an uncontrolled crashed. What would happen if a later CoInitialize caused a DLL to load at that address? Maybe that's improbable thanks to ASLR, but it still feels uncontrolled and undefined to me.

> For what its worth, turns out the CRT option that I was alluding to has already shipped:

Oh, nice! So, would a solution then be to just add this .lib to the link options? E.G. add them to the windows-rs build.rs?

**Comment by kennykerr** [comment]
Yes, I'm not exactly sure whether or how we could do this automatically but if there's a way to do that in the Windows crate I'd be keen to explore it.

**Comment by riverar** [comment]
I can reproduce as-is after elevating the process.

The `AppXDeploymentClient` image is indeed unloaded, causing havoc later with a pointer to invalid memory (non-deterministically) as you indicated. The exception bubbles up and is eventually handled by the [structured exception handling](https://docs.microsoft.com/en-us/cpp/cpp/structured-exception-handling-c-cpp?view=msvc-160&WT.mc_id=WD-MVP-5002756) (SEH) exception filter installed by the runtime.

It's well-known that these handlers exist. Looking at [what undefined behavior (UB) refers to](https://doc.rust-lang.org/nomicon/what-unsafe-does.html#:~:text=undefined%20behavior%20is%20pretty%20limited%20in%20scope%20in%20rust), I'm not sure your crash is UB. Instead, I'd argue the program, and its dependencies, are _incorrect_ in that it called an API that expected things and those things vanished. An exception was thrown and the runtime aborted, eventually calling `ntdll!NtTerminateProcess`. Observe that [this behavior is also documented as safe](https://doc.rust-lang.org/nomicon/what-unsafe-does.html#:~:text=abort%20the%20program).

But I can also appreciate that the API is likely written in an unsafe languages, and Rust safety principles would suggest calling functions written in such a language is unsafe.

I can also appreciate the scary dangling pointer deference and understand, as you pointed out, that it can point to memory that could be re-used and marked executable (which is much more likely on x86 targets where ASLR has very limited space to work with). 

There's lots to noodle through here. That Unsafe Code Guidelines (UCG) Working Group (WG) link you provided looks great, thanks.

Regardless of where we land on _that_ discussion, I think the actionable takeaways are:
* Investigate spinning up a multithreaded apartment (MTA) for devs, to reduce developer pain/incident count
* Continue to spread the message: never call `CoUninitialize`

> Oh, nice! So, would a solution then be to just add this .lib to the link options? E.G. add them to the windows-rs build.rs?

`rustc-link-arg` stabilized recently so that might work. The Rust compiler has some runtime linkage issues though that may complicate matters (e.g. hardcoded to msvcrt.lib). Let us know!

**Comment by bs2kbs2k** [comment]
> a pointer to invalid memory

Maybe I'm getting this wrong, but any reference to uninitialized memory is UB.

**Comment by ljtpetersen** [comment]
> - Continue to spread the message: never call `CoUninitialize`.

I have a question regarding this, and I'm not sure where to ask it. I'm currently writing a safe library that requires COM. However, I am not sure what to do with the initialization. This thread, and especially that quote, tell me that I should initialize COM if not already done, and then never call `CoUninitialize`. However, this contradicts the approach that my other research online has indicated, including [MSN's documentation regarding `CoUninitialize`](https://learn.microsoft.com/en-us/windows/win32/api/combaseapi/nf-combaseapi-couninitialize). What should I do?

**Comment by riverar** [comment]
@ljtpetersen I'd ignore the docs and never call CoUninitialize.

**Comment by tim-weis** [comment]
> There's no downside, the OS will handle cleanup when your process ends.

Automatic cleanup on process termination is well understood. But COM initialization is per thread. While I'm unaware of any (official) information on whether the COM runtime also performs cleanup on thread termination, I still feel uneasy about following the advice.

Is there any documentation of the guarantees implemented by the COM runtime regarding thread termination?

**Comment by riverar** [comment]
Not aware of any official docs myself, but `ole32!DllMain` will get called with a `DLL_THREAD_DETACH` and it will get the opportunity to perform cleanup (e.g. call `ole32!CleanupTlsOleState`).

**Comment by lucatrv** [comment]
> Is there any documentation of the guarantees implemented by the COM runtime regarding thread termination?

Look at the official [`windows`](https://crates.io/crates/windows) crate samples, neither one is calling the [`CoUninitialize`](https://learn.microsoft.com/en-us/windows/win32/api/combaseapi/nf-combaseapi-couninitialize) function.

**Comment by tim-weis** [comment]
> > Is there any documentation of the guarantees implemented by the COM runtime regarding thread termination?
> 
> Look at the official [`windows`](https://crates.io/crates/windows) crate samples, neither one calls the [`CoUninitialize`](https://learn.microsoft.com/en-us/windows/win32/api/combaseapi/nf-combaseapi-couninitialize) function.

Samples aren't specifications (they are informational, non-binding, and will [disagree](https://learn.microsoft.com/en-us/windows/win32/wmisdk/accessing-performance-data-in-c--), too). I was interested in documentation, but to this day the *official* [documentation](https://learn.microsoft.com/en-us/windows/win32/api/combaseapi/nf-combaseapi-couninitialize) reads:

> A thread must call `CoUninitialize` once for each successful call it has made to the [`CoInitialize`](https://learn.microsoft.com/en-us/windows/desktop/api/objbase/nf-objbase-coinitialize) or [`CoInitializeEx`](https://learn.microsoft.com/en-us/windows/desktop/api/combaseapi/nf-combaseapi-coinitializeex) function \[...\].

**Comment by lucatrv** [comment]
> I was interested in documentation, but to this day the _official_ [documentation](https://learn.microsoft.com/en-us/windows/win32/api/combaseapi/nf-combaseapi-couninitialize) reads

Yes, but that documentation is not specific to Rust and the [`windows`](https://crates.io/crates/windows) crate.

**Comment by tim-weis** [comment]
> > I was interested in documentation, but to this day the _official_ [documentation](https://learn.microsoft.com/en-us/windows/win32/api/combaseapi/nf-combaseapi-couninitialize) reads
> 
> Yes, but that documentation is not specific to Rust and the [`windows`](https://crates.io/crates/windows) crate.

Neither Rust nor the `windows` crate can change the rules. The contract for any given interface spelled out in the documentation applies to any programming language.

**Comment by lucatrv** [comment]
@tim-weis, if you believe that the sample programs provided with the [windows](https://crates.io/crates/windows) crate do not comply to the official [documentation](https://learn.microsoft.com/en-us/windows/win32/api/combaseapi/nf-combaseapi-couninitialize), you may want to open an issue here, or submit a PR with the required corrections. The official developers should reply to that.

**Comment by riverar** [comment]
These samples demonstrate specific concepts, patterns, or implementations, rather than adhering to every best practice. For full guidance, definitely check out the public documentation!

**Comment by clouds56** [comment]
I'm not sure if I should create a new issue or discuss here. I have something a little different, but related to CoInitializeEx here.
```rust
#[macro_use] extern crate tracing;
use windows::{Graphics::Capture::GraphicsCaptureSession, Win32::{Foundation::S_FALSE, System::WinRT::{RoInitialize, RoUninitialize, RO_INIT_MULTITHREADED}}};

fn init_com() {
  match unsafe { RoInitialize(RO_INIT_MULTITHREADED) } {
    Ok(_) => info!("COM initialized"),
    Err(e) => {
      if e.code() == S_FALSE {
        info!("COM already initialized");
        // Already initialized
      } else {
        panic!("Failed to initialize COM: {:?}", e);
      }
    }
  }
}

fn main() {
  tracing_subscriber::fmt()
    .with_max_level(tracing::Level::DEBUG)
    .try_init().ok();
  std::thread::spawn(|| {
    init_com();
    GraphicsCaptureSession::IsSupported().unwrap();
    // uninit_com();
  }).join().unwrap();
  init_com();
  GraphicsCaptureSession::IsSupported().unwrap(); // (exit code: 0xc0000005, STATUS_ACCESS_VIOLATION) here
  info!("done");
}
```
I init the com interface twice, but the second init_com doesn't work. I think it is related to `windows_core::imp::FactoryCache` here, is it caching some function pointer that is already unloaded?
Shall we invalid all the cache after all COM instance unloaded?

Note: no `CoUninitialize` is explicitly called here.
Comes from https://github.com/NiiightmareXD/windows-capture/issues/87

**Comment by clouds56** [comment]
Thanks for quick reply, I notice there's `clear_factory_cache()` in your example, I think it might be exactly what's missing here.

I think there's nothing wrong with my code here. If I replace the cached `IsSupported` and call from factory manually, everything goes just well.
```rust
use windows::{Graphics::Capture::{GraphicsCaptureSession, IGraphicsCaptureSessionStatics}, Win32::{System::WinRT::{RoInitialize, RO_INIT_MULTITHREADED}}};


fn init_com() {
  unsafe { RoInitialize(RO_INIT_MULTITHREADED) }.unwrap();
}

fn call_is_supported_without_cache() {
  let factory = <windows::core::imp::FactoryCache<GraphicsCaptureSession, IGraphicsCaptureSessionStatics>>::new();
  factory.call(|this| unsafe {
    let mut result__ = core::mem::zeroed();
    (windows::core::Interface::vtable(this).IsSupported)(windows::core::Interface::as_raw(this), &mut result__).map(|| result__)
  }).unwrap();
}

fn main() {
  std::thread::spawn(|| {
    init_com();
    call_is_supported_without_cache(); // GraphicsCaptureSession::IsSupported().unwrap();
  }).join().unwrap();
  init_com();
  call_is_supported_without_cache(); // GraphicsCaptureSession::IsSupported().unwrap();
  println!("done");
}
```
IIRC, It is the `windows-rs` that caching wrong function pointer here, isn't it?

**Comment by kennykerr** [comment]
The factory cache I wrote for C++/WinRT supports unloading for complicated reasons required by the OS but introduces a performance hit as well as general instability if not handled very _very_ carefully, which is to say its virtually impossible to do reliably. For Rust, I opted not to support unloading for these reasons. I even ended up porting Rust's behavior to C++ as it was preferable in most cases. So in most respects the caching is identical. https://github.com/microsoft/cppwinrt/pull/1031

Both of your examples have subtle bugs related to apartment management. The fact that the code does not crash under light testing does not make it right. 🙃 Due to the nature of COM lifetime, there's just no reliable way to tear down an apartment for all but the most trivial example without terminating the process. A long-lived MTA is the only recommended approach. 

Finally, neither C++ nor Rust for Windows provides per-apartment caching because it is incredibly slow and unreliable. 

Anyway, the `factory` function is provided by the `windows-core` crate if you prefer not to use the factory cache. The `IGraphicsCaptureSessionStatics` interface is however exclusive to the runtime class so you'd need to generate it yourself with `windows-bindgen` as the `windows` crate does not provide it outside the context of the `GraphicsCaptureSession` class.
